File: clutter/lib/models/polygon.dart
import 'coordinate.dart';
import 'types.dart';

class Polygon {
  final int id;
  final String? descriptor;
  final List<Coordinate> coordinates;

  const Polygon({
    required this.id,
    this.descriptor,
    required this.coordinates,
  });

  factory Polygon.fromJson(JsonObject json) {
    if (json
        case {
          'id': int id,
          'descriptor': String? descriptor,
          'coordinates': List coordinates,
        }) {
      return Polygon(
        id: id,
        descriptor: descriptor,
        coordinates:
            coordinates.map((coord) => Coordinate.fromJson(coord)).toList(),
      );
    }
    throw JSONException("Polygon", json);
  }

  JsonObject toJson() {
    return {
      'id': id,
      'descriptor': descriptor,
      'coordinates': coordinates.map((coord) => coord.toJson()).toList(),
    };
  }
}

File: go-server/models/polygon.go
package models

type Polygon struct {
	ID          *int         `json:"id,omitzero"`
	Descriptor  *string      `json:"descriptor,omitzero"`
	Coordinates []Coordinate `json:"coordinates,omitzero"`
	Area        *Area        `json:"area,omitzero"`
}

func (p Polygon) IsZero() bool {
	return p.ID == nil && p.Descriptor == nil && len(p.Coordinates) == 0 && p.Area == nil
}
